### PR TITLE
[HttpClient] Add `HttpClientInterface::withOptions()`

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

I've been thinking about this method for a few months already.
We miss a way to configure an HTTP client in a generic way.
This is useful when eg building an API client as this allows configuring default options once for a consumer, eg in the constructor.

```php
$this->client = $client->withOptions(['base_uri' => 'https://...']);

// [...]

$response = $this->client->request('GET', '/relative-url');
```

### CODE REVIEW

File: src/Symfony/Contracts/HttpClient/HttpClientInterface.php [hunk 1]
```diff
@@ -19,6 +19,8 @@
  *
  * @see HttpClientTestCase for a reference test suite
  *
+ * @method static withOptions(array $options) Returns a new instance of the client with new default options
```
Comment (Nyholm) [inline src/Symfony/Contracts/HttpClient/HttpClientInterface.php:22]: This is not pretty. :/

**Review by Nyholm — APPROVED** [review]
I like the implementation. I think it is a good thing that the `withOptions()` is implemented in a way to keep the HttpClient immutable. 

I can also see that this is useful in third party code.

File: src/Symfony/Contracts/HttpClient/HttpClientInterface.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/HttpClientInterface.php:22]: `DebugClassLoader` will use this to trigger deprecations when implems don't have the method.

**Review by chalasr — APPROVED** [review]

**Review by welcoMattic — APPROVED** [review]

File: src/Symfony/Contracts/HttpClient/HttpClientInterface.php (hunk 1 above)
Comment (emancypage) [reply src/Symfony/Contracts/HttpClient/HttpClientInterface.php:22]: Why withOptions() is described as static? Is it intended? Everywhere it is implemented as normal method.

Comment (nicolas-grekas) [reply src/Symfony/Contracts/HttpClient/HttpClientInterface.php:22]: `static` is the return type. See #42214 or #42358 for previous discussions on the topic.

**Pull request merged**